### PR TITLE
Feature 32731 tool arg

### LINKED ISSUE #32731: Tool Argument Validation Test Cases
Given that a user can potentially supply incorrect or incomplete data in the tool arguments which describe method parameters,
When a tool is defined in an erroneous way,
Then an error should be given to describe the problem and the app will not start

**_Investigate the following edge cases:_** 

- duplicate parameter names
- difference in case sensitivity
- arg name missing
- arg name value is blank

### PR DESCRIPTION
**Pull request by martindrozdz**
- [X] I have considered the risk of behavior change or other zero migration impact (https://github.com/OpenLiberty/open-liberty/wiki/Behavior-Changes).
- [X] If this PR fixes an Issue, the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN" (verify `release bug` label if applicable: https://github.com/OpenLiberty/open-liberty/wiki/Open-Liberty-Conventions).
- [X] If this PR resolves an external Known Issue (including APARS), the description includes "Fixes #FILLMEIN" or "Resolves #FILLMEIN".

Resolves #32731

### CODE REVIEW

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 1]
```diff
@@ -50,25 +58,52 @@ public static ToolMetadata createFrom(Tool annotation, Bean<?> bean, AnnotatedMe
...
+        return new ToolMetadata(annotation, bean, method, getArgumentMap(method), getSpecialArgumentList(method), name, title, description, qualifiedName, businessExceptions);
+    }
+
+    public static ToolMetadata createFrom(Tool annotation, Map<String, ArgumentMetadata> arguments, List<SpecialArgumentMetadata> specialArguments) {
+
+        // used for unit Tests that pre-populate argumentData and create Tools within the tests
+        String title = annotation.title().isEmpty() ? null : annotation.title();
+        String qualifiedName = ""; // for unit tests we do not output qualified names as these are for the CDI aftervalidation deploymentProblem events
+        return new ToolMetadata(annotation, null, null, arguments, specialArguments, annotation.name(), title, annotation.description(), qualifiedName, null);
     }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Move this utility constructor to the test that uses it.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 2]
```diff
@@ -50,25 +58,52 @@ public static ToolMetadata createFrom(Tool annotation, Bean<?> bean, AnnotatedMe
...
             ToolArg pInfo = p.getAnnotation(ToolArg.class);
             if (pInfo != null) {
-                ArgumentMetadata pData = new ArgumentMetadata(p.getBaseType(), p.getPosition(), pInfo.description(), pInfo.required());
-                if (pInfo.name().equals(Tool.ELEMENT_NAME)) {
-                    result.put(method.getJavaMember().getName(), pData);
-                } else {
-                    result.put(pInfo.name(), pData);
-                }
+                setArgumentData(method, result, p, pInfo, argNamesDuplicationCheck);
             }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Since none of your validation conditions prevent you from creating the `ArgumentMetadata` or `ToolMetadata`, and some of the validation checks involve multiple arguments, it will be simpler to add a validate method to `ToolMetadata` which does the validation checks by examining all of the metadata objects.

This means you have all the data available when you validate and separates the validation logic from the construction logic.

File: dev/io.openliberty.mcp.internal_fat/build.gradle [hunk 3]
```diff
@@ -18,4 +18,7 @@ configurations.requiredLibs {
 dependencies {
   requiredLibs 'org.json:json:20080701',
                'org.skyscreamer:jsonassert:2.0-rc1'
+}
+compileJava {
+    options.compilerArgs << '-parameters'
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/build.gradle:25]: Add a comment to say why we're doing this.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+import componenttest.custom.junit.runner.FATRunner;
+import componenttest.topology.impl.LibertyServer;
+import componenttest.topology.utils.FATServletClient;
+import io.openliberty.mcp.internal.fat.tool.toolArgEdgeCasesApp.ToolArgEdgeCasesApp;
+
+/**
+ *
+ */
+@RunWith(FATRunner.class)
+public class ToolArgEdgeCasesTest extends FATServletClient {
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: I don't like calling this "Edge Cases Test", it's more validation, right? Also, why is it a separate test from DeploymentProblemTest? All of these tests also cause deployment problems, right?

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+    }
+
+    @Test
+    public void testToolArgumentEdgeCases() throws Exception {
+        String line1 = server.waitForStringInLog("Blank arguments found in MCP Tool", 5 * 1000);
+        String line2 = server.waitForStringInLog("Duplicate arguments found in MCP Tool", 5 * 1000);
+        boolean foundExpectedErrorinLogs = (line1 != null) || (line2 != null);
+
+        assertTrue("Expected a deployment error due to incorrect Tool Arguments", foundExpectedErrorinLogs);
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: Don't provide a timeout here - usually the default timeout is appropriate and has been tuned to work reliably in the build.

Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: Please use the following strategy for searching for deployment error log messages:

- Wait for either the startApp or stopApp message
- assert that the message found is the stopApp message
- search for all validation log messages
- check that the log messages found are the ones that are expected

This has the following benefits:
- minimises the amount of time spent waiting in the case that the expected messages do not appear
- will fail if unexpected validation messages are emitted

Comment (Azquelt) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: If we have a lot of tests, we'll end up doing these steps a lot and we could wrap them up into a utility. See the Fault Tolerance [ValidationTest](https://github.com/OpenLiberty/open-liberty/blob/integration/dev/com.ibm.ws.microprofile.faulttolerance.cdi_fat/fat/src/com/ibm/websphere/microprofile/faulttolerance_fat/validation/ValidationTest.java) for an example.

**Review by Azquelt — CHANGES_REQUESTED** [review]

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 1 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: It is used in two places: -
io.openliberty.mcp.internal.test.MessageParsingTest
io.openliberty.mcp.internal.test.MCPServerToolsListTest
I created a UtilityClass for it in io.openliberty.mcp.internal.test, but it was only one method.
In hindsight I should have kept the method in the test package. Do you agree?

Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Yes

Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: I know it can feel odd having a class with only one method, but that is sometimes the right choice.

The more important thing here is that we don't want to include test code in the product if we can avoid it. (Sometimes we can't avoid it, but here we can.)

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 2 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Good advice, it felt like the method was doing too much which was making the code overly complex and difficult to read. The separation of concerns has brought me some peace of mind as I wasn't immediately sure what the problem was.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java (hunk 5 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: @Azquelt I have added the requested changes to the PR.
Please review.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/DeploymentProblemTest.java [hunk 6]
```diff
@@ -40,39 +35,45 @@ public class DeploymentProblemTest extends FATServletClient {
...
-                          "id": 1,
-                          "method": "tools/list",
-                          "params": {
-                            "cursor": "optional-cursor-value"
-                          }
-                        }
-                        """;
+    public void testAppFailedToStart() throws Exception {
+        assertTrue(ExpectedAppFailureUtil.appHasFailed(server));
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/DeploymentProblemTest.java]: Unfortunately, the test ordering in JUnit is undefined. This test is not guaranteed to run first.

In this case you could move the checking into the `setup` method.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java (hunk 5 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ToolArgEdgeCasesTest.java]: @Azquelt - as requested: these are the error messages output.

```
java.lang.Exception: Blank arguments found in MCP Tool:
Tool: io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.ToolArgValidationTest.argNameisBlank()
Tool: io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.ToolArgValidationTest.argNameisBlankVariant()
```

```
java.lang.Exception: Duplicate arguments found in MCP Tool:
Tool: io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.ToolArgValidationTest.duplicateParam() -  Argument: arg
Tool: io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.ToolArgValidationTest.duplicateParamVariant() -  Argument: arg
```

```
java.lang.Exception: More than one MCP tool has the same name: 
Tool: bob -- Methods found:
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.bob()
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.duplicateBob()
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest2.duplicateBob()
Tool: echo -- Methods found:
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.duplicateEcho()
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.echo()
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest2.duplicateEcho()
    - io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest2.echo()
```

File: dev/io.openliberty.mcp.internal_fat/build.gradle [hunk 7]
```diff
@@ -18,4 +18,8 @@ configurations.requiredLibs {
 dependencies {
   requiredLibs 'org.json:json:20080701',
                'org.skyscreamer:jsonassert:2.0-rc1'
+}
+/* added below to allow for method parmaters to be available for MCP Tool Server*/
+compileJava {
+    options.compilerArgs << '-parameters'
 }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/build.gradle]:
```suggestion
compileJava {
    // Allow us to test getting the name of a tool argument directly from the Java parameter name
    options.compilerArgs << '-parameters'
}
```

We want to remind ourselves why we're doing this, not just what the argument does.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+        WebArchive war = ShrinkWrap.create(WebArchive.class, warFileName).addPackage(packageName);
+        ShrinkHelper.exportDropinAppToServer(server, war, DISABLE_VALIDATION, SERVER_ONLY);
+        server.startServer();
+    }
+
+    public static boolean appHasFailed(LibertyServer server) throws Exception {
+        // This needs to be called as the first test method in the test FAT class
+        server.waitForStringInLog(APP_STARTED_CODE + "|" + APP_START_FAILED_CODE); // Wait until either the app start or fails to start
+        return !server.findStringsInLogs(APP_START_FAILED_CODE).isEmpty();
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: When you have a helper method that returns a result, it's easy to forget to assert the result.

While you've done it correctly in your test, if you always expect to assert on the result of the method, it's usually better to have the method itself do the assert and provide a useful message

It would also be a good idea to check the application name. While in your test there's only one app, you might use this later in a situation where there's more than one app.

Guidance for how to call the method should be added as Javadoc, not inline comments. The IDE can show you the Javadoc when you're writing the method call.

It's not possible to call something in "the first test method", JUnit is not deterministic in the order in which tests are run.

Lastly, we can just check the log line that was returned, we don't need to re-search the log file.

```suggestion
    /**
     * Asserts that a log message is found indicating that an app has failed to start.
     * This should usually be called before checking for any other validation messages.
     * 
     * @param server the liberty server
     * @param appName the app name to search for
     */
    public static boolean assertAppHasFailed(LibertyServer server, String appName) throws Exception {
        String logLine = server.waitForStringInLog(APP_STARTED_CODE + "|" + APP_START_FAILED_CODE + ".*" + Pattern.quote(appName)); // Wait until either the app starts or fails to start
        assertNotNull("No app start messages found for " + appName, logLine);
        Matcher matcher = Pattern.compile(APP_START_FAILED_CODE).matcher(logLine);
        assertTrue("App " + appName + " did not fail to start. Found log: " + logLine, matcher.find());
    }
```

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java [hunk 9]
```diff
@@ -0,0 +1,51 @@
...
+
+    public static void findAndAssertAllExpectedErrorsInLogs(String errMsg, List<String> expectedErrorList, LibertyServer server) throws Exception {
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: Document what should be supplied for `errMsg` vs. `expectedErrorList`

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java [hunk 10]
```diff
@@ -0,0 +1,51 @@
...
+        for (String err : expectedErrorList) {
+            assertTrue(errMsg + " [" + err + "] expected and not found", !server.findStringsInLogs(err).isEmpty());
+        }
+    }
+
+    public static void stopServerIgnoringErrorMessages(LibertyServer server) throws Exception {
+        server.stopServer(APP_START_FAILED_CODE);
+    }
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: This method is probably not useful. Each test needs to decide which error messages it should ignore.

While at the moment we're only matching the app start failed message, when we do translation every error message will have its own error code and we'll need to ignore any of them that are expected.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java [hunk 11]
```diff
@@ -0,0 +1,51 @@
...
+import org.jboss.shrinkwrap.api.spec.WebArchive;
+
+import com.ibm.websphere.simplicity.ShrinkHelper;
+
+import componenttest.topology.impl.LibertyServer;
+
+public class ExpectedAppFailureUtil {
+
+    private static String APP_STARTED_CODE = "CWWKZ000[13]I";
+    private static String APP_START_FAILED_CODE = "CWWKZ000[24]E";
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: Make these `final`.

You could also make them public so tests can use them when you get rid of `stopServerIgnoringErrorMessages`.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/DeploymentProblemTest.java [hunk 12]
```diff
@@ -40,39 +35,45 @@ public class DeploymentProblemTest extends FATServletClient {
...
 
-        boolean deploymentErrorOccured = false;
+    @Test
+    public void testDuplicateToolDeploymentError() throws Exception {
+        List<String> expectedErrorList = List.of("io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.bob()",
+                                                 "io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.duplicateBob()",
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/DeploymentProblemTest.java]: You've removed a check that was present in the original test and I think is easy to add back in here.

```suggestion
        List<String> expectedErrorList = List.of("More than one MCP tool has the same name",
                                                 "io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.bob()",
                                                 "io.openliberty.mcp.internal.fat.tool.deploymentErrorApps.DuplicateToolErrorTest.duplicateBob()",
```

File: dev/io.openliberty.mcp.internal_fat/publish/servers/mcp-server/bootstrap.properties [hunk 13]
```diff
@@ -1,2 +1,2 @@
 bootstrap.include=../testports.properties
-com.ibm.ws.logging.trace.specification=*=info=:MCP=all
+#com.ibm.ws.logging.trace.specification=*=info=:MCP=all
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal_fat/publish/servers/mcp-server/bootstrap.properties]: Need to revert this change.

When a failure occurs in the build, the trace is the only information we have to diagnose what happened.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpCdiExtension.java [hunk 14]
```diff
@@ -37,13 +39,53 @@ void registerTools(@Observes ProcessManagedBean<?> pmb) {
         for (AnnotatedMethod<?> m : type.getMethods()) {
             Tool toolAnnotation = m.getAnnotation(Tool.class);
             if (toolAnnotation != null) {
-                registerTool(toolAnnotation, pmb.getBean(), m, javaClass.getName());
+                String qualifiedName = javaClass.getName() + "." + m.getJavaMember().getName() + "()";
+                registerTool(toolAnnotation, pmb.getBean(), m, qualifiedName);
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/McpCdiExtension.java]:
```suggestion
                String qualifiedName = javaClass.getName() + "." + m.getJavaMember().getName();
                registerTool(toolAnnotation, pmb.getBean(), m, qualifiedName);
```

Don't add brackets like this when referring to a method.

Either include correct information about method parameters, or just use the name.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 15]
```diff
@@ -29,10 +31,16 @@
 public record ToolMetadata(Tool annotation, Bean<?> bean, AnnotatedMethod<?> method,
                            Map<String, ArgumentMetadata> arguments,
                            List<SpecialArgumentMetadata> specialArguments,
-                           String name, String title, String description,
+                           String name, String title, String description, String qualifiedName,
                            List<Class<? extends Throwable>> businessExceptions) {
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: The qualified name is only used when we want to give an error about the tool and I think it can be computed from the `bean` and `method` data, so we don't need to store it in the `ToolMetadata`, we can just compute it when we need it.

You could add a method to `ToolMetadata` which computes the qualified name.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 16]
```diff
@@ -41,34 +49,44 @@ public record SpecialArgumentMetadata(SpecialArgumentType type, int index) {}
...
+            return ToolArgStatus.PASSED_VALIDATION;
+        } else if (toolArgName.isBlank()) {
+            return ToolArgStatus.BLANK;
+        }
+
+        boolean duplicateArgumentExists = !argNamesDuplicationCheck.add(toolArgName);
+        ToolArgStatus validationStatus = duplicateArgumentExists ? ToolArgStatus.DUPLICATE : ToolArgStatus.PASSED_VALIDATION;
+        return validationStatus;
+    }
+
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: I still think this would be better done afterwards.

- Create the tool and argument metadata without doing any validation.
  - Do not store `ToolArgStatus`
- Validate the created metadata objects afterwards from `McpCdiExtension`.

**Review by Azquelt — CHANGES_REQUESTED** [review]

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java (hunk 10 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: I've added this in otherwise the build fails. I'm not entirely sure why to be honest.

Comment (Azquelt) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: Rather than having this method, include the `server.stopServer(APP_START_FAILED_CODE);` call in each test class that needs it.

Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: OK, from my understanding I believe that it just ignoring the APP_START_FAILED_CODE error message.
I am only using it in the tearDown() method.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 15 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: I get the qualifiedName from the javaClass which I can only get from the AnnotatedType in McpCdiExtension.registerTools():

```
AnnotatedType<?> type = pmb.getAnnotatedBeanClass();
Class<?> javaClass = type.getJavaClass();
Then I use javaClass.getName() to get the class information.
```

I don't see any other way of getting this information outside of registerTools.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 16 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: There is a problem here in that on line 72 if toolArgname is already in the result HashMap then it will be overridden.

File: dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java (hunk 8 above)
Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal_fat/fat/src/io/openliberty/mcp/internal/fat/tool/ExpectedAppFailureUtil.java]: I'm not really sure how the interaction with App name should work.
In this example below (from io.openliberty.microprofile.reactive.messaging.internal_fat/fat/src/io/openliberty/microprofile/reactive/messaging/fat/validation/ValidationTests.java) the appName is simply the test method name:

```
@Test
    public void testIncomingMethodWithoutUpstream() throws Exception {
        String testAppName = "IncomingMethodWithoutUpstreamTest";
        AppValidator.validateAppOn(server)
                        .withAppName(testAppName)
                        .withClass(IncomingMethodWithoutUpstream.class)
                        .failsWith("Wiring error\\(s\\) detected in application")
                        .failsWith("CWMRX1100E:.*" + testAppName)
                        .failsWith("- SubscriberMethod\\{method:'[^']*', incoming:'" + IncomingMethodWithoutUpstream.CHANNEL + "'\\} has no upstream")
                        .run();
//        Suppressed: io.smallrye.reactive.messaging.providers.wiring.OpenGraphException: Some components are not connected to either downstream consumers or upstream producers:
//        - SubscriberMethod{method:'io.openliberty.microprofile.reactive.messaging.fat.validation.IncomingMethodWithoutUpstream#badMethod', incoming:'IncomingMethodWithoutUpstream'} has no upstream
    }
```

In the AppValidator class the appName is being used to create a filename for the war file (See below):
```
            if (appName != null) {
                archiveName = appName + ".war";
            } else {
                archiveName = "testApp-" + (++appCount) + ".war";
            }

            try {
                WebArchive app
```

I've pushed the fork, and can look at this task in the next issue if that is alright?

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 17]
```diff
@@ -58,11 +54,11 @@ private static Map<String, ArgumentMetadata> getArgumentMap(AnnotatedMethod<?> m
         for (AnnotatedParameter<?> p : method.getParameters()) {
             ToolArg pInfo = p.getAnnotation(ToolArg.class);
             if (pInfo != null) {
-                ArgumentMetadata pData = new ArgumentMetadata(p.getBaseType(), p.getPosition(), pInfo.description(), pInfo.required());
-                if (pInfo.name().equals(Tool.ELEMENT_NAME)) {
-                    result.put(method.getJavaMember().getName(), pData);
-                } else {
-                    result.put(pInfo.name(), pData);
+                String toolArgName = (pInfo.name().equals(Tool.ELEMENT_NAME)) ? p.getJavaParameter().getName() : pInfo.name(); // p.getJavaParameter().getName() needs java compiler -parameter flag to work
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]:
```suggestion
                String toolArgName = (pInfo.name().equals(ToolArg.ELEMENT_NAME)) ? p.getJavaParameter().getName() : pInfo.name(); // p.getJavaParameter().getName() needs java compiler -parameter flag to work
```

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java [hunk 18]
```diff
@@ -58,11 +54,11 @@ private static Map<String, ArgumentMetadata> getArgumentMap(AnnotatedMethod<?> m
...
+                ArgumentMetadata ArgInfo = new ArgumentMetadata(p.getBaseType(), p.getPosition(), pInfo.description(), pInfo.required(), false);
+                boolean isDuplicateArg = result.put(toolArgName, ArgInfo) != null;
+                if (!toolArgName.isBlank() && isDuplicateArg) {
+                    result.put(toolArgName, new ArgumentMetadata(p.getBaseType(), p.getPosition(), pInfo.description(), pInfo.required(), isDuplicateArg));
```
Comment (Azquelt) [inline dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]:
```suggestion
                boolean isDuplicateArg = result.contains(toolArgName);
                ArgumentMetadata argInfo = new ArgumentMetadata(p.getBaseType(), p.getPosition(), pInfo.description(), pInfo.required(), isDuplicateArg);
                result.put(toolArgName, argInfo);
```

It's more straightforward to just check if the argument name is a duplicate before adding it to the map, rather than adding it, and then updating it if it was a duplicate.

`ArgInfo` -> `argInfo` because it's a variable name.

I'm not sure what the purpose of the if `!toolArgName.isBlank()` check was.

Comment (martindrozdz) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Aah, yes. Good catch. I could have just done a get as well. 
I'll use boolean:
`isDuplicateArg = result.containsKey(toolArgName); `
as contains() doesn't exist.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 16 above)
Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: We agreed this is indeed a problem.

File: dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java (hunk 18 above)
Comment (Azquelt) [reply dev/io.openliberty.mcp.internal/src/io/openliberty/mcp/internal/ToolMetadata.java]: Ah yes, `containsKey` is what you want :+1:

**Review by Azquelt — APPROVED** [review]
